util/encoding/encoding.go

```diff
@@ -444,3 +453,255 @@ func DecodeBytes(b []byte) ([]byte, []byte) {
+				i++
+			} else if format[i] == '-' {
+				ascending = false
+				i++
+			}
+		}
+
+		// Process width specifier (either blank, "32" or "64").
+		width := 0
+		if i+1 < end {
+			if format[i] == '3' && format[i+1] == '2' {
+				width = 32
+				i += 2
+			} else if format[i] == '6' && format[i+1] == '4' {
+				width = 64
+				i += 2
+			}
+		}
+
+		// Process verb.
+		if i < end {
+			switch format[i] {
+			case 'd':
+				if ascending {
+					b = EncodeVarint(b, args[0].(int64))
+				} else {
+					b = EncodeVarintDecreasing(b, args[0].(int64))
+				}
+			case 'u':
+				if ascending {
+					if width == 0 {
+						b = EncodeUvarint(b, args[0].(uint64))
+					} else if width == 32 {
+						b = EncodeUint32(b, args[0].(uint32))
+					} else {
+						b = EncodeUint64(b, args[0].(uint64))
+					}
+				} else {
+					if width == 0 {
+						b = EncodeUvarintDecreasing(b, args[0].(uint64))
+					} else if width == 32 {
+						b = EncodeUint32Decreasing(b, args[0].(uint32))
+					} else {
+						b = EncodeUint64Decreasing(b, args[0].(uint64))
+					}
+				}
+			case 's':
+				var arg []byte
+				if s, ok := args[0].(string); ok {
+					// Treat the string as a []byte to avoid an allocation when
+					// converting to a string. This is kosher only because the
+					// string header and slice header match on the first two
+					// fields and we know EncodeBytes{,Decreasing} does not keep
+					// a reference to the value it encodes.
+					arg = *(*[]byte)(unsafe.Pointer(&s))
+				} else {
+					arg = args[0].([]byte)
+				}
+				if ascending {
+					b = EncodeBytes(b, arg)
+				} else {
+					b = EncodeBytesDecreasing(b, arg)
+				}
+			default:
+				panic("unknown format verb")
+			}
+			i++
+			args = args[1:]
+		}
+	}
+	return b
+}
+
+// DecodeKey decodes values from a byte slice according to the format
+// string. Returns the remainder of the byte slice if it was not
+// completely consumed by the format string. See EncodeKey for details
+// of the format specifiers.
+//
+// The variadic arguments must be pointers to the types specified by
+// the format string. For example, '%d" requires a '*int64' argument.
+func DecodeKey(b []byte, format string, args ...interface{}) []byte {
+	i, end := 0, len(format)
+	for i < end && format[i] != '%' {
+		if len(b) == 0 || b[i] != format[i] {
+			panic("format mismatch")
+		}
+		i++
+	}
+	if i > 0 {
+		b = b[i:]
+	}
+
+	for i < end {
+		if format[i] != '%' {
+			panic("invalid format string: " + format)
+		}
+		i++
+
+		// Process ascending flag (either "+" or "-").
+		ascending := true
+		if i < end {
+			if format[i] == '+' {
+				i++
+			} else if format[i] == '-' {
+				ascending = false
+				i++
+			}
+		}
+
+		// Process width specifier (either blank, "32" or "64").
+		width := 0
+		if i+1 < end {
+			if format[i] == '3' && format[i+1] == '2' {
+				width = 32
+				i += 2
+			} else if format[i] == '6' && format[i+1] == '4' {
+				width = 64
+				i += 2
+			}
+		}
+
+		// Process verb.
+		if i < end {
```

Comment:
It looks like this fails silently if the string ends with an unterminated specifier; this `if` should probably have an `else` that panics.